docs/proposals/0000-support-interchangeable-build-and-run-implementations.md

```diff
@@ -0,0 +1,248 @@
+outside the `korifi` codebase, so that these flows can be customized and implemented in different ways. To support this,
+we will add the new `BuildWorkload` and `RunWorkload` resources, which contain all information necessary to execute
+their respective processes. The `korifi-controllers` will create these resources, and separate (interchangeable)
+components will reconcile them to build images and run long-running processes. We will provide default implementations
+of these reconcilers that work the same way Korifi currently works (via kpack Images and statefulsets,
+respectively), but operators can choose to deploy alternatives that implement these flows differently.
+
+# Motivation
+[motivation]: #motivation
+
+The build and run implementations are currently baked into korifi-controllers and cannot be changed. While the current
+implementation of these flows may be sufficient for most use cases, there are many reasons operators may want to swap in
+alternate logic. Some examples:
+
+- Using `Deployment`, knative, or other alternative resources for run workloads
+- Signing container images in the build step, and requiring signed images in the run step
+- Security scanning container images during the build and/or run steps
+- Applying conventions to pod spec, such as labels, annotations, volume mounts, etc.
+
+We could accomplish this flexibility with a smaller change: splitting out the current reconcilers for `CFBuild` and
+`CFProcess` into their own repositories, so they can be replaced with alternatives. The issue with this change is that
+the existing reconcilers use information from many resources and an alternative implementation would have to duplicate
+this complex logic. For example, to create the LRPs that currently generate the run step's stateful set, the `CFProcess`
+reconciler uses `CFProcess`, `CFApp`, `CFPackage`, `CFServiceBinding`, and `Secret` resources.
+
+The new `BuildWorkload` and `RunWorkload` solve this problem by placing all necessary information on a single resource.
+As a consequence, the build and run controllers can focus on implementing their respective flows without having to fetch
+and correlate multiple resources. These resources also provide a standard way to report status back to
+the `korifi-controllers` without directly modifying Korifi domain resources.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Here are the various resources and components involved:
+
+- `BuildWorkload` - a new Custom Resource Definition, with fields that contain all necessary information to execute the
+  build step for a korifi workload. This resource contains information from `CFBuild`, `CFPackage`, `CFServiceBinding`,
+  and the App's `Secret` resources. This CRD may exist in a new, separate API Group to indicate its status as an
+  external interface.
+
+- `RunWorkload` - a new Custom Resource Definition, with fields that contain all necessary information to execute the
```

Comment:
I think @gcapizzi is also thinking through something to model CF Tasks. I'm wondering could be a separate case of this one or a dedicated `TaskWorkload` resource.

Reply:
I think it would be clearer if we had a dedicated `TaskWorkload` for that, since it would be reconciled differently.

Reply:
Yeah, and that would then cover the three types of Diego tasks (staging, LRP, and task) Cloud Controller makes today so it's unlikely we'd need even more.